### ISSUE TITLE
`Signal::set` needs to panic instead of silently failing on double-borrow

### ISSUE THREAD

**Issue by ltfschoen** [maciejhirsz/kobold#56]
In this draft PR https://github.com/maciejhirsz/kobold/pull/55, I was trying to see if I could use the `get` function implementation of `Hook` https://docs.rs/kobold/latest/kobold/stateful/struct.Hook.html#method.get just for fun.
It states that its used to _"Get the value of state if state implements Copy"_. I used a state that is just a struct `State` that has a `my_state` property with a `bool` type so it can implement the `Copy` trait, since if the `State` was more complex and had properties of types that don't automatically implement the `Copy` trait (i.e. `Vec`, `String`, `Range`, `Box`) then you get errors like the following, which were unnecesassarily complicated for me to resolve when I just wanted to a simple example of using the `Hook`'s `get` function :
```
error[E0204]: the trait `Copy` may not be implemented for this type
  --> examples/invoice/src/state.rs:33:32
   |
33 | #[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
   |                                ^^^^
...
39 |     pub qr_code: String,
   |     ------------------- this field does not implement `Copy`
   |
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0204]: the trait `Copy` may not be implemented for this type
  --> examples/invoice/src/state.rs:41:32
   |
41 | #[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
   |                                ^^^^
42 | pub struct Entry {
43 |     pub description: String,
   |     ----------------------- this field does not implement `Copy`
   |
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0204]: the trait `Copy` may not be implemented for this type
  --> examples/invoice/src/state.rs:47:32
   |
47 | #[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
   |                                ^^^^
...
50 |     pub columns: Vec<Text>,
   |     ---------------------- this field does not implement `Copy`
51 |     pub rows: Vec<Vec<Text>>,
   |     ------------------------ this field does not implement `Copy`
   |
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0204]: the trait `Copy` may not be implemented for this type
  --> examples/invoice/src/state.rs:54:32
   |
54 | #[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
   |                                ^^^^
55 | pub enum Text {
56 |     Insitu(Range<usize>),
   |            ------------ this field does not implement `Copy`
57 |     Owned(Box<str>),
   |           -------- this field does not implement `Copy`
   |
   = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0204]: the trait `Copy` may not be implemented for this type
   --> examples/invoice/src/state.rs:204:32
    |
204 | #[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
    |                                ^^^^
205 | pub struct TextSource {
206 |     pub source: String,
    |     ------------------ this field does not implement `Copy`
```

So my first thought was, is the intention for developers whose `State` has properties that don't automatically implement the `Copy` trait to try to get them to implement the `Copy` trait if they want to use the `Hook`'s public `get` function (even though it might require changes to the Kobold library itself)? Or is the intention for it to just be used by developers with simple projects where the `State` has properties that automatically implement the `Copy` trait?

Then when I was playing with trying different parts of the docs, I tried to use `stateful::Signal` public function `.set`  https://docs.rs/kobold/latest/kobold/stateful/struct.Signal.html#method.set, which says it's used to  _"Replace the entire state with a new value and trigger an update."_, but when I tried using it to change the state, it actually didn't change the state at all as expected, i thought it would have set the `my_state` value to `true`. See screenshot below of UI output:

<img width="1331" alt="Screen Shot 2023-04-04 at 12 24 31 pm" src="https://user-images.githubusercontent.com/6226175/229672945-e580e59f-4144-47e4-9e5d-e6c8cd9d108b.png">

Note: I also tried using `update`, but that doesn't appear to have worked either for some reason even though I thought I'd been using it previously without issues

**Comment by maciejhirsz** [comment]
The `set` and `get` are working _somewhat_ similar to `std::cell::Cell` here. The only time you really need the `get` method is if your entire state is something like `count: &Hook<u32>` and you want to get the `u32` by value. Since `Hook<T>` implements `Deref<Target = T>` you can copy a value out of it by doing `**count` (one `*` to deref the `&`, the other to deref the `Hook`), `count.get()` is just a convenience method to make that slightly cleaner and to show the intent clearer. 

You can't implement `Copy` on a type that has `String`s or `Vec`s in it as that would violate ownership rules: `String` is a smart pointer that owns the memory it points to, when you _clone_ it a new allocation is made, the content is copied to the new allocation, and then you get a new `String` with a new pointer. This is different from, say, `&str` which is a shared reference so the pointer can be just copied on stack with no issues.

If you have a `state: &Hook<State>` where `State` has a `my_state` field that has a bool in it, you can get that bool just by invoking `state.my_state` and `Deref` will take care of the rest. This derefencing is free after optimizations (the references `&Hook<State>` and dereferenced `&State` are the same pointer to the compiler).

As to the `Signal::set`: do you have a link to where you use it? Chances are it fails silently because you're trying to set state while already having a mutable reference to it. This needs to panic (at least in debug mode), I just haven't gotten to it yet.

**Comment by ltfschoen** [comment]
> As to the Signal::set: do you have a link to where you use it?

yes i created this draft PR to illustrate how i encountered the Signal::set issue https://github.com/maciejhirsz/kobold/pull/55/files

i mock the state here [here](https://github.com/maciejhirsz/kobold/pull/55/files#diff-7bd1d43f7fc04cf42330101a8dddc73633120511683183f1aea5f88a2823232dR32) so `state.my_state` has a value `false` when injected into the `Hello` component [here](https://github.com/maciejhirsz/kobold/pull/55/files#diff-d3595088ee41ba12a376817a06a7b217cf194070c22a982a98b375635d2b471fR10).

and i pass a prop `new_state` with value `true` into [here](https://github.com/maciejhirsz/kobold/pull/55/files#diff-d3595088ee41ba12a376817a06a7b217cf194070c22a982a98b375635d2b471fR34)

it correctly gets `state.my_state` as `false` [here](https://github.com/maciejhirsz/kobold/pull/55/files#diff-d3595088ee41ba12a376817a06a7b217cf194070c22a982a98b375635d2b471fR34),

but then i run `signal.update(|state| state.toggle());`, which should toggle the value of `state.my_state` to be `true` [here](https://github.com/maciejhirsz/kobold/pull/55/files#diff-7bd1d43f7fc04cf42330101a8dddc73633120511683183f1aea5f88a2823232dR62), but that didn't work

then i also tried `signal.set(State { my_state: new_state });` [here](https://github.com/maciejhirsz/kobold/pull/55/files#diff-7bd1d43f7fc04cf42330101a8dddc73633120511683183f1aea5f88a2823232dR62), which should set the value of `state.my_state` to be `true`, but that didn't work either.

**Comment by maciejhirsz** [comment]
Fixed in #64.
